File: javaScript/monk_and_match_making.js
process.stdin.resume();
process.stdin.setEncoding('utf-8');

let input = '';
process.stdin.on('data', function(chunk) {
    input += chunk;
});

process.stdin.on('end', function() {
    const lines = input.trim().split('\n');
    let idx = 0;
    
    // Read the string
    const s = lines[idx++];
    const n = s.length;
    
    // Base and modulus for rolling hash
    const base = 911382323n;
    const mod = 1000000007n;
    
    // Precompute powers
    const powArr = new Array(n + 2);
    powArr[0] = 1n;
    for (let i = 1; i <= n; i++) {
        powArr[i] = (powArr[i - 1] * base) % mod;
    }
    
    // Compute prefix hashes
    const hashArr = new Array(n + 2);
    hashArr[0] = 0n;
    for (let i = 0; i < n; i++) {
        const val = BigInt(s.charCodeAt(i) - 96);
        hashArr[i + 1] = (hashArr[i] * base + val) % mod;
    }
    
    // Read number of queries
    const q = parseInt(lines[idx++]);
    
    // Process each query
    const output = [];
    for (let i = 0; i < q; i++) {
        const [L1, R1, L2, R2] = lines[idx++].split(' ').map(Number);
        
        // Check if lengths are different
        if ((R1 - L1) !== (R2 - L2)) {
            output.push("No");
            continue;
        }
        
        // Calculate hash for first substring
        let h1 = (hashArr[R1] - (hashArr[L1 - 1] * powArr[R1 - L1 + 1]) % mod + mod) % mod;
        
        // Calculate hash for second substring
        let h2 = (hashArr[R2] - (hashArr[L2 - 1] * powArr[R2 - L2 + 1]) % mod + mod) % mod;
        
        if (h1 === h2) {
            output.push("Yes");
        } else {
            output.push("No");
        }
    }
    
    console.log(output.join('\n'));
});

File: python/monk_and_match_making.py
import sys

def solve():
    input = sys.stdin.readline
    
    # Read the string
    s = input().strip()
    n = len(s)
    
    # Base and modulus for rolling hash
    base = 911382323
    mod = 1000000007
    
    # Precompute powers
    pow_arr = [1] * (n + 2)
    for i in range(1, n + 1):
        pow_arr[i] = (pow_arr[i - 1] * base) % mod
    
    # Compute prefix hashes
    hash_arr = [0] * (n + 2)
    for i in range(n):
        val = ord(s[i]) - ord('a') + 1
        hash_arr[i + 1] = (hash_arr[i] * base + val) % mod
    
    # Read number of queries
    q = int(input())
    
    # Process each query
    for _ in range(q):
        L1, R1, L2, R2 = map(int, input().split())
        
        # Check if lengths are different
        if (R1 - L1) != (R2 - L2):
            print("No")
            continue
        
        # Calculate hash for first substring
        h1 = (hash_arr[R1] - (hash_arr[L1 - 1] * pow_arr[R1 - L1 + 1]) % mod + mod) % mod
        
        # Calculate hash for second substring
        h2 = (hash_arr[R2] - (hash_arr[L2 - 1] * pow_arr[R2 - L2 + 1]) % mod + mod) % mod
        
        if h1 == h2:
            print("Yes")
        else:
            print("No")

if __name__ == "__main__":
    solve()